fairlearn/reductions/exponentiated_gradient/exponentiated_gradient.py

```diff
@@ -0,0 +1,267 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+"""
+This module implements the Lagrangian reduction of fair binary
+classification to standard binary classification.
+
+FUNCTIONS
+exponentiated_gradient -- optimize accuracy subject to fairness constraints
+"""
+
+from __future__ import print_function
+
+import numpy as np
+import pandas as pd
+import functools
+from fairlearn.reductions import moments
+from fairlearn.reductions import ReductionsEstimator
+from ._constants import _ACCURACY_MUL, _REGRET_CHECK_START_T, _REGRET_CHECK_INCREASE_T, \
+    _SHRINK_REGRET, _SHRINK_ETA, _MIN_T, _RUN_LP_STEP, _PRECISION, _INDENTATION
+from ._lagrangian import _Lagrangian
+from collections import namedtuple
+
+print = functools.partial(print, flush=True)
+
+
+def _mean_pred(X, hs, weights):
+    """Return a weighted average of predictions produced by classifiers in hs"""
+    pred = pd.DataFrame()
+    for t in range(len(hs)):
+        pred[t] = hs[t](X)
+    return pred[weights.index].dot(weights)
+
+
+# TODO: decide whether we want a structured object instead of the namedtuple
+column_names = "best_classifier best_gap classifiers weights last_t best_t n_oracle_calls"
+expgrad_result = namedtuple("ExpgradResult", column_names)
+
+
+class ExponentiatedGradientResult:
+    def __init__(self, best_classifier, best_gap, classifiers, weights, last_t, best_t,
+                 n_oracle_calls):
+        self._best_classifier = best_classifier
+        self._best_gap = best_gap
+        self._classifiers = classifiers
+        self._weights = weights
+        self._last_t = last_t
+        self._best_t = best_t
+        self._n_oracle_calls = n_oracle_calls
+
+    def _predict_with_best_classifier(self, X):
+        return self._best_classifier(X)
+
+    def _asdict(self):
+        return {
+            "best_classifier": self._best_classifier,
+            "best_gap": self._best_gap,
+            "classifiers": self._classifiers,
+            "weights": self._weights,
+            "last_t": self._last_t,
+            "best_t": self._best_t,
+            "n_oracle_calls": self._n_oracle_calls
+        }
+
+
+class ExponentiatedGradient(ReductionsEstimator):
+    def __init__(self, estimator, disparity_metric, quality_metric, eps=0.01):
+        self._estimator = estimator
+        self._disparity_metric = disparity_metric
+        # TODO use quality metric
+        self._quality_metric = quality_metric
+        self._eps = eps
+        self._best_classifier = None
+        self._classifiers = None
```

Comment:
What happened to T, nu and eta_mul? Shouldn't they be set in the constructor too?

Reply:
Absolutely! Just pushed really hard for this to work, and the others just worked by default. Will come!